### ISSUE TITLE
Native async callbacks

### ISSUE THREAD

**Issue by vaind** [objectbox/objectbox-dart#143]
To implement any kind of async callbacks from C to Dart, we have to jump through some hoops, see the following issues:
* https://github.com/dart-lang/sdk/issues/37022
* https://github.com/dart-lang/sdk/issues/35766
* https://github.com/flutter/flutter/issues/63255#issuecomment-671216406

In short, native callbacks must be executed on the "right" dart isolate, during its event loop.  
This issue currently affects observers (#142) and is a prerequisite for `SyncClient` listeners (#140).

From my understanding, getting this to work is quite a bit involved because we would need to create some custom C code that would call directly into dart APIs to inject the callback into the event loop (forgive if this is inaccurate, still learning about the workaround). AFAICT this would require that we either ship a different version of the objectbox-c or we create an additional one we ship alongside the standard one used by all other languages...

**Comment by vaind** [comment]
These are async callbacks that should be supported:

```c
/// Callback for obx_observe()
/// @param user_data user data given to obx_observe()
/// @param type_ids array of object type IDs that had changes
/// @param type_ids_count number of IDs of type_ids
typedef void obx_observer(void* user_data, const obx_schema_id* type_ids, size_t type_ids_count);

/// Callback for obx_observe_single_type()
typedef void obx_observer_single_type(void* user_data);

/// Called when connection is established
/// @param arg is a pass-through argument passed to the called API
typedef void OBX_sync_listener_connect(void* arg);

/// Called when connection is closed/lost
/// @param arg is a pass-through argument passed to the called API
typedef void OBX_sync_listener_disconnect(void* arg);

/// Called on successful login
/// @param arg is a pass-through argument passed to the called API
typedef void OBX_sync_listener_login(void* arg);

/// Called on a login failure
/// @param arg is a pass-through argument passed to the called API
/// @param code error code indicating why the login failed
typedef void OBX_sync_listener_login_failure(void* arg, OBXSyncCode code);

/// Called when synchronization is complete
/// @param arg is a pass-through argument passed to the called API
typedef void OBX_sync_listener_complete(void* arg);

/// Called with fine grained sync changes (IDs of put and removed entities)
/// @param arg is a pass-through argument passed to the called API
typedef void OBX_sync_listener_change(void* arg, const OBX_sync_change_array* changes);
```

There's also data visitor but that one is synchronous (called directly from dart-called C function)

**Comment by vaind** [comment]
#145 - Managed to get this working with custom c-api endpoints, as described in the link in the original issue description.
Waiting for the new c-api to be released (or however we want to distribute those dart-specific native symbols).

**Comment by vaind** [comment]
Implemented by #145
